### PR TITLE
feat: added Information screen

### PR DESCRIPTION
**Pull request by Shukriii**
# DCMAW-7797: Information screen

This PR creates a screen with an image, title, body, (optional) footnote, primary button and an (optional) secondary button. The weight of the image can be changed allowing this screen to be reused for various information screens.

image

# Checklist

## Before raising your pull request:
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with ticket ID and a short description about the feature or update
      i.e. _DCMAW-222: Added ReadID SDK to iOS app_
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

- [ ] Met all accessibility requirements?
    - [x] Checked dynamic type sizes are applied
    - [x] Checked VoiceOver can navigate your new code
    - [ ] Checked a user can navigate only using a keyboard around your new code 

## Before merging your pull request:
- [ ] Ensure that the code coverage and SonarCloud checks have passed
- [ ] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [ ] Ran the app to ensure that no regressions have been caused by changes during code review.
- [ ] Targeted the correct branch; `develop`, `release` or `main`

### CODE REVIEW

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift [hunk 1]
```diff
@@ -63,8 +64,7 @@ enum Screens: String, CaseIterable {
                                                                screenView: {}, dismissAction: {})
             return InstructionsWithImageViewController(viewModel: viewModel)
         case .gdsModalInfoView:
-            let viewModel = MockModalInfoViewModel()
-            let view = ModalInfoViewController(viewModel: viewModel)
+            let view = ModalInfoViewController(viewModel: MockModalInfoViewModel())
```
Comment (Shukriii) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:67]: Refactored here and line 95, as swift lint was complaining that the function was over 50 lines.

File: GDSCommon-Demo.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 2]
```diff
@@ -5,8 +5,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/nalexn/ViewInspector",
       "state" : {
-        "revision" : "a1422d4749ccf7f32b1d14a9ec19ec9a0b0fd337",
-        "version" : "0.9.9"
+        "revision" : "67319287749b83f39dcc2f20edd520c610c012fd",
```
[the remark this answers is not in the document]
Comment (DylanRJ) [reply GDSCommon-Demo.xcworkspace/xcshareddata/swiftpm/Package.resolved:8]: I think that should be fine, shouldn't be a breaking change

**Comment by robinphillips** [comment]
> Simulator Screenshot - iPhone 15 Pro - 2024-01-30 at 12 38 59 Something I've noticed here with scaling dynamic type is the fact that the screen content gets totally hidden when the size gets up to maximum. Might be one to check with the BAs/Designers on the ticket 🙂

We could set a maximum text size for it. See here:
https://useyourloaf.com/blog/restricting-dynamic-type-sizes/

Its iOS 15+ but that would cover the vast majority of our users :)

**Comment by Shukriii** [comment]
Changed `footnoteLabel.maximumContentSizeCategory = .accessibilityLarge` to `.accessibilityMedium` (to show more of the content), looks like this on the maximum zoom on a SE.

image

**Comment by robinphillips** [comment]
> Changed `footnoteLabel.maximumContentSizeCategory = .accessibilityLarge` to `.accessibilityMedium` (to show more of the content), looks like this on the maximum zoom on a SE.
> image

I feel like perhaps we should have a discussion with the designers about where the ideal compromise is here? For example:
- maybe the mock footnote you have is longer than is ever likely to be used and therefore the maximum dynamic type size could increase
- perhaps we should look to modify the maximum dynamic type size for the buttons 

It feels like there should be a discussion with the designers about what the ideal compromise is.

File: GDSCommon-Demo/GDSCommon-DemoTests/InstructionsWithImageViewControllerTests.swift [hunk 3]
```diff
@@ -53,13 +53,13 @@ final class InstructionsWithImageViewControllerTests: XCTestCase {
 }
 
 extension InstructionsWithImageViewControllerTests {
-    func testDidAppear() {
+    func test_didAppear() {
```
Comment (robinphillips) [inline GDSCommon-Demo/GDSCommon-DemoTests/InstructionsWithImageViewControllerTests.swift:56]: Please can you revert the changes that are not related to the ticket?

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationController.swift [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import UIKit
+
+/// View controller for `GDSInformation` screen
+///     - `informationImage` (type: `String`)
+///     - `titleLabel` (type: `UILabel`)
+///     - `bodyLabel` (type: `UILabel`)
+///     - `footnoteLabel`  (type: `UILabel`)
+///     - `primaryButton`  (type: ``RoundedButton`` inherits from ``SecondaryButton``)
+///     - `secondaryButton`  (type: ``SecondaryButton`` inherits from ``UIButton``)
+public final class GDSInformationController: BaseViewController, TitledViewController {
```
Comment (benmckillop) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationController.swift:10]: I think this should be `GDSInformationViewController` rather than `GDSInformationController` :) Also the same for the name of the test files.

Comment (Shukriii) [reply Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationController.swift:10]: thanks Ben making this change now :)

File: GDSCommon-Demo/GDSCommon-DemoTests/GDSInformationControllerTests.swift [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    func test_secondaryButtonAction() throws {
+        XCTAssertFalse(secondaryButton)
+        try sut.informationSecondaryButton.sendActions(for: .touchUpInside)
+        XCTAssertTrue(secondaryButton)
+    }
+    
+    func test_didAppear() throws {
+        XCTAssertFalse(viewDidAppear)
+        sut.beginAppearanceTransition(true, animated: false)
+        sut.viewDidAppear(false)
```
Comment (DylanRJ) [inline GDSCommon-Demo/GDSCommon-DemoTests/GDSInformationControllerTests.swift:107]: I don't think we should be calling these lifecycle methods directly if we can avoid them, the `.beginAppearanceTransition(_:, animated:)` and `.endAppearanceTransition()` should be sufficient for managing the screens displaying programmatically.
Maybe take them out and see how the tests perform 🙂

File: GDSCommon-Demo/GDSCommon-DemoTests/GDSInformationControllerTests.swift [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    let rightBarButtonTitle: GDSLocalisedString? = "right bar button"
+    let backButtonIsHidden: Bool = false
+    let appearAction: () -> Void
+    let dismissAction: () -> Void
+    
+    init(primaryButtonAction: @escaping () -> Void,
+         secondaryButtonAction: @escaping () -> Void,
+         appearAction: @escaping () -> Void,
+         dismissAction: @escaping () -> Void
+    ) {
```
Comment (DylanRJ) [inline GDSCommon-Demo/GDSCommon-DemoTests/GDSInformationControllerTests.swift:55]: I think this can hug the argument on the previous line 🙂

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationViewController.swift [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    @IBOutlet private var informationImage: UIImageView! {
+        didSet {
+            let font = UIFont(style: .largeTitle, weight: viewModel.imageWeight ?? .semibold)
+            let configuration = UIImage.SymbolConfiguration(font: font, scale: .large)
+            informationImage.image = UIImage(systemName: viewModel.image, withConfiguration: configuration)
+            informationImage.tintColor = viewModel.imageColour ?? .gdsPrimary
+            informationImage.accessibilityIdentifier = "information-image"
+            
+            NSLayoutConstraint.activate([
+                informationImage.heightAnchor.constraint(greaterThanOrEqualToConstant: viewModel.imageHeightConstraint ?? 55)
```
Comment (robinphillips) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationViewController.swift:34]: Did you look at figuring out the calculation for transforming the `viewModel.imageHeightConstraint` so that devs inputting a height for the image get the height they expect?

Comment (Shukriii) [reply Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInformation/GDSInformationViewController.swift]: Currently compensating the padding by adding 11, and scaling set to large

**Review by DylanRJ — APPROVED** [review]

**Review by mcgarj — APPROVED** [review]
